### PR TITLE
Files-App Versions-Tab

### LINKED ISSUE #639: File Versions sidebar
## Current version side bar:

screenshot from 2019-02-08 11-21-38

## Apis to use:
https://owncloud.github.io/js-owncloud-client/FilesVersions.html

## Implementation proposal
- simple list (v-list)
- show version timestamp + size + download icon + icon to restore a version
- implement the sidebar within the files app - no new app

### PR DESCRIPTION
**Pull request by scramb**
## Description

This PR will introduce the Versions-Tab to the Files-App.

## Related Issue

- Refs #639 
- Refs #579 

## Screenshots:
screenshot_2019-02-14 owncloud phoenix 1

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Open tasks:

- [ ] Reload versions if deselecting from multiple items
- [x] add file size
- [x] add revert and download functionallity
- [ ] fix revert and download functionallity

### CODE REVIEW

**Comment by scramb** [comment]
@DeepDiver1975 am i missing some DAV properties to get the fileSize from version files? and what is the VersionID and am i missing a DAV property here aswell?

**Comment by DeepDiver1975** [comment]
> am i missing some DAV properties to get the fileSize from version files?

this is getcontentlength



> and what is the VersionID and am i missing a DAV property here aswell?

the version id is the last path element

/meta/999/v/1550130960 - version id  is 1550130960 (yes - looks like a timestamp which it is in this implementation ;-) )

File: apps/files/src/components/FileInfoVersions.vue [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      })
+    },
+    revertVersion (file, restorePath = '/') {
+      this.$client.fileVersions.restoreFileVersion(this.currentFile, file.fileInfo['{DAV:}getetag'], restorePath).then((res) => {
+        console.log(res)
+      })
+    },
+    downloadVersion (file) {
+      console.log(file)
+      this.$client.fileVersions.getFileVersionContents(this.currentFile, file.fileInfo['{DAV:}getetag']).then((res) => {
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileInfoVersions.vue]: Maybe we better use fetch here and pipe the response object into an blob url ... see the other download implementation

File: apps/files/src/components/FileInfoVersions.vue [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  },
+  methods: {
+    getFileVersions () {
+      this.$client.fileVersions.listVersions(this.currentFile).then((res) => {
+        this.versions = res
+      })
+    },
+    revertVersion (file, restorePath = '/') {
+      this.$client.fileVersions.restoreFileVersion(this.currentFile, file.fileInfo['{DAV:}getetag'], restorePath).then((res) => {
+        console.log(res)
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileInfoVersions.vue]: - [ ] version list needs to be reloaded
- [ ] file list needs to be reloaded/adjusted as well - size, modified date and some day preview are changing after restoiring of a version

**Comment by DeepDiver1975** [comment]
@scramb I rebased this PR ....

File: apps/files/src/components/FileInfoVersions.vue (hunk 1 above)
Comment (scramb) [reply apps/files/src/components/FileInfoVersions.vue]: For now i cant get the Path of the version file, only it's content; Not sure how to implement fetch here the same way.

Comment (DeepDiver1975) [reply apps/files/src/components/FileInfoVersions.vue]: you get the path using this method: https://github.com/owncloud/js-owncloud-client/blob/8d27b316b2dec02c37d024b858b5820ef7d98c50/src/fileVersionManagement.js#L123

**Comment by DeepDiver1975** [comment]
conflicting :see_no_evil:

File: apps/files/src/components/FileInfoVersions.vue [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  computed: {
+    ...mapGetters('Files', ['selectedFiles']),
+    ...mapGetters(['getToken']),
+    currentFile () {
+      return this.selectedFiles[0]
+    }
+  },
+  methods: {
+    restorePath (file) {
+      let filename = this.currentFile.path.split('.', 1)[0]
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileInfoVersions.vue]: such logic will fail for
- filenames without extension
- filenames with more than 2 dots - e.g. foo.tar.gz

File: apps/files/src/components/FileInfoVersions.vue [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      let split = this.currentFile.path.split('.')
+      let extension = split[split.length - 1]
+      return `${filename}_v${this.currentVersionId(file)}.${extension}`
+    },
+    currentVersionId (file) {
+      let etag = file.name.split('/')
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileInfoVersions.vue]: name should already hold the correct value - there is something else borked

File: apps/files/src/components/FileInfoVersions.vue [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        this.versions = res
+      })
+    },
+    revertVersion (file, restorePath = '/') {
+      this.$client.fileVersions.restoreFileVersion(this.currentFile.id, this.currentVersionId(file), restorePath).then(() => {
+        this.$emit('reload')
+      })
+    },
+    downloadVersion (file) {
+      console.log(file)
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileInfoVersions.vue]: debug left over

File: apps/files/src/components/FileactionsTab.vue [hunk 6]
```diff
@@ -33,6 +50,11 @@ export default {
         name: appId
       })
     },
+    downloadFiles (file) {
+      // TODO fix missing bearer Token / broken $client
+      console.log('js-owncloud-client issue - Bearer Token missing', file)
+      this.downloadFile(file)
```
Comment (DeepDiver1975) [inline apps/files/src/components/FileactionsTab.vue]: where is this code change coming from - should not have been touched ...

File: apps/files/src/components/FilesApp.vue [hunk 7]
```diff
@@ -23,10 +23,16 @@ import { mapActions, mapGetters, mapState } from 'vuex'
 const _includes = require('lodash/includes')
 
 const davProperties = [
+  '{http://owncloud.org/ns}permissions',
```
Comment (DeepDiver1975) [inline apps/files/src/components/FilesApp.vue]: something wrong on rebase - this was part of a different change

File: apps/files/src/components/FilesTopBar.vue [hunk 8]
```diff
@@ -79,6 +79,9 @@
   </v-card>
 </v-menu>
       </template>
+      <template slot="action_new" v-if="!this.canUpload">
```
Comment (DeepDiver1975) [inline apps/files/src/components/FilesTopBar.vue]: already in master - rebase :boom:

File: apps/files/src/components/FileactionsTab.vue (hunk 6 above)
Comment (scramb) [reply apps/files/src/components/FileactionsTab.vue]: This was the addition to #579. Any idea why Bearer is undefined?

Comment (DeepDiver1975) [reply apps/files/src/components/FileactionsTab.vue]: different topic -> different PR - don't mix changes

Comment (DeepDiver1975) [reply apps/files/src/components/FileactionsTab.vue]: undefined bearer could be related to the initialization process - we had this already in the search somewhere ....

File: apps/files/src/components/FileInfoVersions.vue (hunk 4 above)
Comment (scramb) [reply apps/files/src/components/FileInfoVersions.vue]: > /meta/999/v/1550130960 - version id is 1550130960 (yes - looks like a timestamp which it is in this implementation ;-) )

name holds  /meta/999/v/1550130960

Comment (DeepDiver1975) [reply apps/files/src/components/FileInfoVersions.vue]: I'll debug this - just a sec ....

**Comment by scramb** [comment]
rebased in #670

**Pull request closed without merging**